File: python/src/onetool/commands/shared/io.py
from __future__ import annotations

from dataclasses import dataclass

from ...execution_policy import check_materialized_byte_limit, format_materialized_limit_message
from ...types import CommandResult, err
from ...utils import format_size, looks_binary, parent_path
from ..core import CommandContext, CommandSpec


@dataclass(frozen=True, slots=True)
class TextInputResult:
    text: str | None = None
    error: CommandResult | None = None


def render_help(spec: CommandSpec) -> str:
    return f"{spec.name}: {spec.summary}\nUsage: {spec.usage}\n\n{spec.details}"


async def read_text_from_file_or_stdin(
    ctx: CommandContext,
    file_path: str | None,
    stdin: bytes,
    command_name: str,
) -> TextInputResult:
    if file_path is not None:
        info = await ctx.vfs.stat(file_path)
        normalized = ctx.vfs.normalize(file_path)
        if not info.exists:
            return TextInputResult(
                error=err(
                    f"{command_name}: file not found: {normalized}. "
                    f"Use: ls / or ls {parent_path(file_path)}"
                )
            )
        if info.is_dir:
            return TextInputResult(
                error=err(f"{command_name}: path is a directory: {normalized}. Use: ls {normalized}")
            )

        limit_error = materialized_limit_error(ctx, command_name, normalized, info.size)
        if limit_error is not None:
            return TextInputResult(error=limit_error)

        raw = await ctx.vfs.read_bytes(file_path)
        if looks_binary(raw):
            return TextInputResult(
                error=err(
                    f"{command_name}: binary file ({format_size(len(raw))}): {normalized}. "
                    f"Use: stat {normalized}"
                )
            )
        return TextInputResult(text=raw.decode("utf-8"))

    if stdin:
        limit_error = materialized_limit_error(ctx, command_name, "stdin", len(stdin))
        if limit_error is not None:
            return TextInputResult(error=limit_error)

        if looks_binary(stdin):
            return TextInputResult(
                error=err(
                    f"{command_name}: stdin is binary ({format_size(len(stdin))}). "
                    "Pipe text only, or save binary data to a file and inspect with stat."
                )
            )
        return TextInputResult(text=stdin.decode("utf-8"))

    return TextInputResult(error=err(f"{command_name}: no input provided. Usage: see 'help {command_name}'"))


def content_from_args_or_stdin(
    args: list[str],
    stdin: bytes,
    command_name: str,
) -> tuple[bytes | None, CommandResult | None]:
    if len(args) > 1:
        return (" ".join(args[1:]).encode("utf-8"), None)
    if stdin:
        return (bytes(stdin), None)
    return (
        None,
        err(f"{command_name}: no content provided. Pass content inline or pipe it via stdin."),
    )


def parse_n_flag(args: list[str], default_value: int) -> tuple[int, list[str], CommandResult | None]:
    if not args:
        return (default_value, [], None)
    if args[0] != "-n":
        return (default_value, args, None)
    if len(args) < 2:
        return (default_value, args, err("missing value for -n"))

    try:
        parsed = int(args[1], 10)
    except ValueError:
        return (default_value, args, err(f"invalid integer for -n: {args[1]}"))
    return (parsed, args[2:], None)


def materialized_limit_error(
    ctx: CommandContext,
    command_name: str,
    subject: str,
    actual_bytes: int,
) -> CommandResult | None:
    exceeded = check_materialized_byte_limit(ctx.execution_policy, actual_bytes)
    if exceeded is None:
        return None
    return err(format_materialized_limit_message(command_name, subject, exceeded.actual, exceeded.limit))

File: src/commands/shared/io.ts
import type { CommandResult } from '../../types.js';
import { err, textDecoder, textEncoder } from '../../types.js';
import { formatSize, looksBinary, parentPath } from '../../utils.js';
import type { CommandContext, CommandSpec } from '../core.js';
import { checkMaterializedByteLimit, formatMaterializedLimitMessage } from '../../execution-policy.js';

export function renderHelp(spec: CommandSpec): string {
  return `${spec.name}: ${spec.summary}\nUsage: ${spec.usage}\n\n${spec.details}`;
}

export type TextInputResult =
  | { text: string; error?: never }
  | { text?: never; error: CommandResult };

export async function readTextFromFileOrStdin(
  ctx: CommandContext,
  filePath: string | undefined,
  stdin: Uint8Array,
  commandName: string,
): Promise<TextInputResult> {
  if (filePath) {
    const info = await ctx.vfs.stat(filePath);
    const normalized = ctx.vfs.normalize(filePath);
    if (!info.exists) {
      return {
        error: err(`${commandName}: file not found: ${normalized}. Use: ls / or ls ${parentPath(filePath)}`),
      };
    }
    if (info.isDir) {
      return {
        error: err(`${commandName}: path is a directory: ${normalized}. Use: ls ${normalized}`),
      };
    }
    const limitError = materializedLimitError(ctx, commandName, normalized, info.size);
    if (limitError !== undefined) {
      return { error: limitError };
    }
    const raw = await ctx.vfs.readBytes(filePath);
    if (looksBinary(raw)) {
      return {
        error: err(
          `${commandName}: binary file (${formatSize(raw.length)}): ${normalized}. Use: stat ${normalized}`,
        ),
      };
    }
    return { text: textDecoder.decode(raw) };
  }

  if (stdin.length > 0) {
    const limitError = materializedLimitError(ctx, commandName, 'stdin', stdin.length);
    if (limitError !== undefined) {
      return { error: limitError };
    }
    if (looksBinary(stdin)) {
      return {
        error: err(
          `${commandName}: stdin is binary (${formatSize(stdin.length)}). ` +
            'Pipe text only, or save binary data to a file and inspect with stat.',
        ),
      };
    }
    return { text: textDecoder.decode(stdin) };
  }

  return {
    error: err(`${commandName}: no input provided. Usage: see 'help ${commandName}'`),
  };
}

export function contentFromArgsOrStdin(
  args: string[],
  stdin: Uint8Array,
  commandName: string,
): { ok: true; value: Uint8Array } | { ok: false; error: CommandResult } {
  if (args.length > 1) {
    return { ok: true, value: textEncoder.encode(args.slice(1).join(' ')) };
  }
  if (stdin.length > 0) {
    return { ok: true, value: stdin };
  }
  return {
    ok: false,
    error: err(`${commandName}: no content provided. Pass content inline or pipe it via stdin.`),
  };
}

export function parseNFlag(
  args: string[],
  defaultValue: number,
): { value: number; remaining: string[]; error?: CommandResult } {
  if (args.length === 0) {
    return { value: defaultValue, remaining: [] };
  }
  if (args[0] !== '-n') {
    return { value: defaultValue, remaining: args };
  }
  if (args.length < 2) {
    return { value: defaultValue, remaining: args, error: err('missing value for -n') };
  }
  const parsed = Number.parseInt(args[1]!, 10);
  if (!Number.isFinite(parsed)) {
    return {
      value: defaultValue,
      remaining: args,
      error: err(`invalid integer for -n: ${args[1]}`),
    };
  }
  return { value: parsed, remaining: args.slice(2) };
}

export function materializedLimitError(
  ctx: CommandContext,
  commandName: string,
  subject: string,
  actualBytes: number,
): CommandResult | undefined {
  const exceeded = checkMaterializedByteLimit(ctx.executionPolicy, actualBytes);
  if (!exceeded) {
    return undefined;
  }

  return err(formatMaterializedLimitMessage(commandName, subject, exceeded.actual, exceeded.limit), {
    exitCode: 1,
  });
}
